cli/command/manifest/push.go

```diff
@@ -42,32 +47,63 @@ type pushRequest struct {
 	insecure      bool
 }
 
+type yamlManifestList struct {
+	Image     string
+	Manifests []yamlManifest
+}
+
+type yamlManifest struct {
+	Image    string
+	Platform manifestlist.PlatformSpec
+}
+
 func newPushListCommand(dockerCli command.Cli) *cobra.Command {
 	opts := pushOpts{}
 
 	cmd := &cobra.Command{
-		Use:   "push [OPTIONS] MANIFEST_LIST",
+		Use:   "push [OPTIONS] [MANIFEST_LIST]",
 		Short: "Push a manifest list to a repository",
-		Args:  cli.ExactArgs(1),
+		Args:  cli.RequiresRangeArgs(0, 1),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			opts.target = args[0]
+			if len(args) == 1 {
+				opts.target = args[0]
+			}
 			return runPush(dockerCli, opts)
 		},
 	}
 
 	flags := cmd.Flags()
-	flags.BoolVarP(&opts.purge, "purge", "p", false, "Remove the local manifest list after push")
-	flags.BoolVar(&opts.insecure, "insecure", false, "Allow push to an insecure registry")
+	flags.BoolVarP(&opts.purge, "purge", "p", false, "remove the local manifest list after push")
+	flags.BoolVar(&opts.insecure, "insecure", false, "allow push to an insecure registry")
+	flags.StringVar(&opts.file, "filename", "", "create, annotate, and push using a yaml file")
 	return cmd
 }
 
 func runPush(dockerCli command.Cli, opts pushOpts) error {
 
+	if opts.target == "" && opts.file == "" {
+		return errors.New("please specify either a file or local manifest name")
+	}
+
+	if opts.file != "" {
+		yamlInput, err := getYamlManifestList(dockerCli, opts.file)
+		if err != nil {
+			return err
+		}
+		if err := createListFromYaml(dockerCli, yamlInput, opts.insecure); err != nil {
+			return err
+		}
+		if err := annotateListFromYaml(dockerCli, yamlInput); err != nil {
+			return err
+		}
```

Comment:
It seems a bit odd to me that `push` would create local files. could this push directly from yaml without creating the files locally?

If we want to support "load defaults from yaml" or something like that we could have `create` accept the same yaml format (either in this PR or a separate one).